### DISCUSSION TITLE
ESP32C3 - GPIO7 fails to Output I2C SDA Signal when `fs_mount()` is invoked

### DISCUSSION THREAD

**Discussion by jumper385** [zephyrproject-rtos/zephyr#62749] [Q&A] [answered]
Hi All!

I have a custom **ESP32C3** board that I want to use to log sensor data to a csv on an **SD-Card**. The **sensors** (MPU6050 and BME280) are connected to the I2C bus. The chip is talking to them via the **sensor API's**. I intend to use the **Filesystem API** to write logs.

## Setup
GPIO6 - SCL
GPIO7 - SDA

## Firmware Flow
The firmware works by: 
1. Initialising I2C (`&i2c0`) and SPI (`&spi2`) threads. Additionally, each thread will have it's events initialised
2. Main thread will fire a mount event to SPI thread where fs_mount is invoked
3. Main thread will emit events to I2C thread to get IMU and Baro data
4. ***Unimplemented***: Log data to CSV

## Problem
Between when i remove step 2 (invoking `fs_mount`), step 3 can successfully run `sensor_sample_fetch` and subsequently `sensor_sample_get()` and `sensor_value_to_float()` from the sensors (wont throw an error). If i keep step 2, step 3 will not be able to successfully run `sensor_sample_fetch` (the return value is wrong). Additionally, the fs_mount wont throw an error and indicates that the SD-Card was mounted correctly.

Sadly, I'm not sure how to setup the ESP32 OpenOCD to step through firmware properly on zephyr to debug. It'd be great if someone could also point out how I could do this. If its easier, I'm happy to run it purely in terminal. `west debug` didn't seem to get me anywhere and information on setting up OpenOCD for ESP32 debugging is quite sparse. **If needed, I'm also happy to write a brief document on how to set it up that could be used for the zephyr docs too!**

## Links
[UWAA Flight Computer](https://github.com/jumper385/uwaa_flight_computer) is the repo i'm currently working in. Its based off the example project. 
[Schematics](https://github.com/jumper385/uwaa_flight_computer/blob/main/design_files/v0.1/Schematics.PDF): Note there are errors however, the pinout hasn't changed. The errors are 1) no pull-up on reset and 2) 32Mhz Clock NOT 40MHz clock. I've fixed these issues with a Jank and spare 40MHz clock.

**Comment by EricNRS** [comment]
The current Zephyr OpenOCD version 0.11 doesn't support ESP32 yet (not sure when v0.12 will be integrated), so you need to build the ESP32 OpenOCD build and then tell west to use that.  See https://docs.zephyrproject.org/latest/boards/riscv/esp32c3_devkitm/doc/index.html#debugging for how to specify the OpenOCD version to use when doing west build which is then used when you do west debug.

To build the ESP32 Fork of OpenOCD, do the usual:
```
mkdir ~/esp32-openocd && cd $_
git clone --recursive https://github.com/espressif/openocd-esp32.git .
sudo apt-get install make libtool pkg-config autoconf automake texinfo libusb-1.0
./bootstrap  
./configure --prefix=$(pwd)/install  
make -j8 && make install
```

I set it up in this case to do the install to the source folder just to keep things together.  You then follow this with the following commands from your Zephyr build directory:
```
export ESP_OPENOCDBASE=~/esp32-openocd/install
west build -b YOUR_BOARD YOUR_APP -- -DOPENOCD=${ESP_OPENOCDBASE}/bin/openocd -DOPENOCD_DEFAULT_PATH=${ESP_OPENOCDBASE}/share/openocd/scripts
west debug
```

As for your I2C and SD card issues it is hard to tell if you have a pin configuration issue or a threading issue.  I would personally try to run the commands in a single thread as a test case to make sure that works.  Also, check the final DT in build/zephyr/zephyr.dts to see if there is anything there that you are not expecting that would cause a pinctrl issue.

**Reply by jumper385** [reply]
Hey @EricNRS

## Test
I did some further testing and it seems to be failing when the FS mounts. If the FS Fails to mount, i'm able to fetch from the sensors. If the FS mounts however, it will fail to fetch. Any idea what this might be? nothing seems to be shared on the zephyr.dts. Do you think this should be changed to an issue?

### Links
I've created a separate branch called [`sd-debugging`](https://github.com/jumper385/uwaa_flight_computer/tree/sd-debugging) that puts everything in the main thread - only using the structs etc from the linked files - no threads.

## OpenOCD
I did get openocd working however, I had to follow the espressif docs and setup their openocd - it was a convoluted process... then west-debug would work... but it woudl tell me it cant count the number of freertos threads. Your method didnt work. For reference, tho, i'm on a Macbook Air M1.

Henry

**Reply by EricNRS** [reply]
This is an open-source project and the Espressif team has limited time, so the best way to get help is to narrow down the issue and provide a patch to reproduce it with one of the sample apps.  Often while narrowing it down, you find the root cause.

For debugging, once you get more than a few threads going, you will end up debugging a lot of things using logging.  The Tracing framework is also useful, but I personally have not tried it yet (it is on my todo list).

Back to your issue, SPI and I2C should be unrelated, so another place to look is to see if you have an electrical issue with your board such that once you enable the SD card there is increased noise or a sag in the 3V3 power rail.  Put an oscilloscope on the 3V3 bus and the I2C lines to see what is happening on the bus.

You can also enable logging for the I2C bus (I2C_LOG_LEVEL_DBG), the I2C sensor you are using, and the SPI driver for the flash chip and see what is happening.  They may point you in the correct direction.

**Reply by rftafas** [reply]
Hi, thanks for your consideration. 

They can also use Core Dump: https://blog.espressif.com/core-dump-a-powerful-tool-for-debugging-programs-in-zephyr-with-esp32-boards-969830fd6cdb

**Reply by jumper385** [reply]
Hi All!

Thanks for the input @rftafas! @EricNRS was correct - the firmware hasnt crashed so i didnt need a dump. I did start reading the logs of the device and found that the firmware was unable to transfer through I2C. 

![image](https://github.com/zephyrproject-rtos/zephyr/assets/17038915/3f91809f-ebbd-4ad1-86bc-936478a381ae)

I probed the 3V3 Lines and found that no voltage sag was present - I even removed a protection diode and operated the line at 3V3 Volts. The same issue was the same. 

I also probed the I2C lines and found something interesting. When `fs_mount()` wasnt invoked, both, the SDA and SCL were operating. When `fs_mount()` was invoked, however, only the SCL line was outputting - the SDA line was quiet.

<img width="1082" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/17038915/81634542-c98f-4ca5-b0e4-4904f17a9fb9">

Interesting aye...

**Reply by jumper385** [reply, marked as answer]
I may have found a solution but I'm not sure if this is something that should be flagged... Perhaps the docs may be enough to tell others and I may have been a bit dumb 😅

## Findings
I probed the lines and found that the MTDO pin used for JTAG debugging was outputting a signal whenever the device would start up. When I switch the pins to a different pin (GPIO10), I am able to use both, SD Card and I2C. I'm not sure why this would stop the system from spitting out data from the I2C lines but i thought. Should this be flagged as a potential problem? 

For reference, SDA was connected to GPIO7 and SCL was connected to GPIO6.

## Waveforms
<img width="1121" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/17038915/80dbe742-5a9a-4d84-aae2-773fda810b4c">
<img width="699" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/17038915/97dfba28-aa5b-4ad2-be4b-62d1c0cd3e7d">

## Multi-Threading Performance
EDIT: I've since tried this with the original firmware which had the SD-Card and I2C peripheral managed by separate threads. The current fix works.

**Reply by rftafas** [reply]
I understood this is a custom ESP32-C3 board, down to chip. First, make sure power and reset timings are correct. If they aren't, unexpected behavior may happen.

I did a sweep look into IDF reports and found some other issues, but all different from your report. You could try burning the e-fuses and **permanently** disable JTAG on pins (burning EFUSE_DIS_PAD_JTAG) to check if that still happens.

Additional:
- ESP32C3 TRM Chap. 4
- https://docs.espressif.com/projects/esp-idf/en/latest/esp32c3/api-reference/system/efuse.html
- https://github.com/espressif/esptool/releases/tag/v3.3

**Reply by EricNRS** [reply]
> MTDO pin used for JTAG debugging was outputting a signal whenever the device would start up.

Nice work!  Looks like you have the problem figured out and just need to nail down the root cause.

Check in the ESP32-C3 Technical Reference Manual - that is probably standard chip boot-up behaviour as it could also be a boot strapping pin or just be known to have a power-up glitch.

If either the SDA or SCL pins have glitchs on them, it is possible that your I2C devices end up in a bad state and require a reset to recover, so that would explain the issue you are seeing from a sensor standpoint.

**Reply by jumper385** [reply]
@rftafas Cheers for that! I can try to do that but not now - I'm running out of boards but will be doing a new revision of my hardware to test. i'll report back in a few weeks. 
@EricNRS you are right - it's to do with the power-up glitch. I suspect @rftafas is right though, blowing the e-fuses would solve this permanently - specifically to eliminate the MTDO and MTCK glitches which are also seen in the plots (REF ESP32C3 Datasheet - Chapter 2.2 Table 2-2).
![image](https://github.com/zephyrproject-rtos/zephyr/assets/17038915/c7afd9a4-a33c-4947-84a7-02fff8eeb98c) 
I do wonder why it does that when `fs_mount` is invoked. I'll do more digging.

**Reply by jumper385** [reply]
should i rename the discussion to JTAG Glitch Causes I2C line failure? Additionally, are there any other experiments you'd like me to run?

**Reply by EricNRS** [reply]
It looks like you've root caused it and the title is probably fine as is what you observed first -- someone else might search of the same thing in the future.

**Reply by rftafas** [reply]
I can't see the time base for the pictures, but it seems that these are far later in the game to be those glitches. I would still check if the 50uS between power and EN are respected (Tstbl).

Furthermore, the initial 'drop' on GPIO6 may be the glitch (albeit, far larger than the 5nS). But the actions on GPIO7 are not ok. Since MTDO should relate to MTCK, I don't think this is action of JTAG. Seems too fast. Seems FSPI is doing something unexpected. Can you check that?

**Reply by jumper385** [reply]
Hi @rftafas 
Apologies for the late one but i've found some time to samplethe glitch. I've added the annotations here. Woudl you know whats happening here?

GPIO6 Initial Glitch on SCL Line
![image](https://github.com/zephyrproject-rtos/zephyr/assets/17038915/0b7e8513-f71b-4d7c-9c02-2058837d8444)

GPIO7 Initial Glitch on SDA Line - NOTE this repeats now that GPIO10 is used as SDA
![image](https://github.com/zephyrproject-rtos/zephyr/assets/17038915/e5732803-e5e9-4437-8a7e-b6e94a3265d7)
![image](https://github.com/zephyrproject-rtos/zephyr/assets/17038915/525876fe-3d47-4b0c-9aa1-8a1b4d3a615a)

**Reply by rftafas** [reply]
First, please, check your circuit for the EN pin. You are missing a capacitor and not respecting the Enable timing. Also, be aware that there is a 3ms hold time on all strapping pins. If they are connected to something else, make sure something else is respecting that.

**Reply by jumper385** [reply]
hmmm okay... I'm currently doing revisions on the board - I'll implement your suggestions and circle back! Cheers @rftafas

**Reply by rftafas** [reply]
BTW, check that power supply too. Those 2 diodes may drop the voltage below minimum of 3.0V. The garbage you see happens after a voltage drop below minimum. Most likely, when invoking a function that causes higher consumption, you make system very unpredictable.

[edit] [Consider reviewing the datasheet](https://www.espressif.com/sites/default/files/documentation/esp32-c3_datasheet_en.pdf), as it seems several requirements haven't been met.
[edit 2] This system will probably face issues with ESPtool as it is now. [Consider reading the throubleshooting](https://docs.espressif.com/projects/esptool/en/latest/esp32/troubleshooting.html).
[edit 3] ESP32-C3 has accumulated Idd = 0,5A, the regulator TLV7033 is specified to 0,3A. Please, fix that too.

**Comment by EricNRS** [comment]
@rftafas Hi Ricardo.  Henry isn't seeing a crash, he is just seeing that I2C sensor reads are failing when SD card (running in SPI mode) file system is successfully mounted.  This is on an ESP32-C3.

**Reply by rftafas** [reply]
I got it. I'm just listing more options, actually. I should have tagged that intention.

**Reply by EricNRS** [reply]
On a similar topic of debug, do you know that state of Zephyr thread awareness in GDB?  OpenOCD v0.12.0 added Zephyr thread awareness, but it doesn't seem to work yet (I have only tried on ESP32-S3 so far).

**Reply by rftafas** [reply]
Its out of topic, but for the sake of not leaving it hanging: it hasn't been tested yet, at least not by us, so I can't provide an status. Zephyr would first need to upgrade to 0.12 (AFAIK it is based on 0.11) and then we would start testing.

**Reply by EricNRS** [reply]
Thanks, I have opened an enhancement request https://github.com/zephyrproject-rtos/zephyr/issues/62791 to add Zephyr RTOS thread awareness to OpenOCD/GDB.

**Reply by jumper385** [reply]
why does #62791 not include the ESP32C3? is that to do with unknowns with the riscv core? I'm just curious at this point...

**Reply by EricNRS** [reply]
Mainly because I have only tested the support on ESP32-S3, although looking at the code, I don't see any support for RISC-V, either.  I have updated the ticket to cover all ESP32-* processors.
